File: src/main/java/com/adven/concordion/extensions/exam/core/utils/EvaluatorValueResolver.java
package com.adven.concordion.extensions.exam.core.utils;

import com.github.jknack.handlebars.ValueResolver;
import org.concordion.api.Evaluator;

import java.util.*;

public enum EvaluatorValueResolver implements ValueResolver {
    INSTANCE;

    @SuppressWarnings({"rawtypes", "unchecked"})
    @Override
    public Object resolve(final Object context, final String name) {
        Object value = null;
        if (context instanceof Evaluator) {
            value = ((Evaluator) context).getVariable("#" + name);
        }
        return value == null ? UNRESOLVED : value;
    }

    @Override
    public Object resolve(final Object context) {
        if (context instanceof Evaluator) {
            return context;
        }
        return UNRESOLVED;
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    @Override
    public Set<Map.Entry<String, Object>> propertySet(final Object context) {
        if (context instanceof Evaluator) {
            throw new UnsupportedOperationException();
        }
        return Collections.emptySet();
    }
}

File: src/main/java/com/adven/concordion/extensions/exam/core/utils/EvaluatorValueResolver.kt
package com.adven.concordion.extensions.exam.core.utils

import com.github.jknack.handlebars.ValueResolver
import ognl.NoSuchPropertyException
import org.concordion.api.Evaluator
import org.concordion.internal.InvalidExpressionException

enum class EvaluatorValueResolver : ValueResolver {
    INSTANCE;

    override fun resolve(context: Any, name: String): Any {
        var value: Any? = null
        if (context is Evaluator) {
            value = context.getVariable("#$name") ?: context.tryEval(name)
        }
        return value ?: ValueResolver.UNRESOLVED
    }

    private fun Evaluator.tryEval(name: String): Any? = try {
        this.evaluate(name)
    } catch (e: InvalidExpressionException) {
        if (e.cause is NoSuchPropertyException) null
        else throw e
    }

    override fun resolve(context: Any): Any = context as? Evaluator ?: ValueResolver.UNRESOLVED

    override fun propertySet(context: Any?): MutableSet<MutableMap.MutableEntry<String, Any>> {
        if (context is Evaluator) {
            throw UnsupportedOperationException()
        }
        return HashSet()
    }
}
